source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,400 @@
+  auto read = std::next(begin);
+  auto write = std::next(begin);
+  auto end = path.end();
+
+  while (read != end) {
+    if (*read == '/') {
+      char prev = *std::prev(write);
+      if (prev == '/') {
+        // Duplicate slash, merge it
+        ++read;
+      } else {
+        // Not a duplicate slash or we aren't configured to merge slashes, copy it
+        *write++ = *read++;
+      }
+    } else {
+      *write++ = *read++;
+    }
+  }
+
+  path.resize(std::distance(begin, write));
+  return PathNormalizationResult::success();
+}
+
+PathNormalizer::PathNormalizationResult
+PathNormalizer::collapseDotSegmentsPass(std::string& path) const {
+  auto begin = path.begin();
+  auto read = std::next(begin);
+  auto write = std::next(begin);
+  auto end = path.end();
+  absl::string_view path_view{path};
+
+  while (read != end) {
+    if (*read == '.') {
+      char prev = *std::prev(write);
+      if (prev == '/') {
+        // attempt to read ahead 2 characters to see if we are in a "./" or "../" segment.
+        const auto dot_segment = path_view.substr(std::distance(begin, read), 3);
+        if (absl::StartsWith(dot_segment, "./") || dot_segment == ".") {
+          // This is a "/./" segment or the path is terminated by "/.", ignore it
+          size_t distance = std::min<size_t>(dot_segment.size(), 2);
+          // Advance the read iterator by 1 if the path ends with "." or 2 if the segment is "./"
+          std::advance(read, distance);
+        } else if (dot_segment == "../" || dot_segment == "..") {
+          // This is a "/../" segment or the path is terminated by "/..", navigate one segment up.
+          // Back up write 1 position to the previous slash to find the previous segment start.
+          auto new_write = findStartOfPreviousSegment(std::prev(write), begin);
+          if (new_write == begin) {
+            // This is an invalid ".." segment, most likely the full path is "/..", which attempts
+            // to go above the root.
+            return {PathNormalizationResult::Action::Reject,
+                    UhvResponseCodeDetail::get().InvalidUrl};
+          }
+
+          // Set the write position to overwrite the previous segment
+          write = new_write;
+          // Advance the read iterator by 2 if the path ends with ".." or 3 if the segment is "../"
+          size_t distance = std::min<size_t>(dot_segment.size(), 3);
+          std::advance(read, distance);
+        } else {
+          *write++ = *read++;
+        }
+      } else {
+        *write++ = *read++;
+      }
+    } else {
+      *write++ = *read++;
+    }
+  }
+
+  path.resize(std::distance(begin, write));
+  return PathNormalizationResult::success();
+}
+
+std::tuple<absl::string_view, absl::string_view>
+PathNormalizer::splitPathAndQueryParams(absl::string_view path_and_query_params) const {
+  // Split on the query (?) or fragment (#) delimiter, whichever one is first.
+  auto delim = path_and_query_params.find_first_of("?#");
```

Comment:
Envoy's [current behavior](https://github.com/envoyproxy/envoy/blob/ce8129cdf640f66b57559ad888f4e6e69203686d/source/common/http/conn_manager_utility.cc#L540) is to reject URI paths with fragment in them. It can be changed using runtime flag.

This really should be a config option to accept, reject or strip fragment from URI path. A significant portions of requests in GCP have # in URI path.

Please add TODO to make behavior with respect to the # in URI configurable.

Reply:
Done. I added a TODO referencing a new issue: #23886